### ISSUE TITLE
Geometry checksum/hash for consistency checks between simulation and reconstruction steps

### ISSUE THREAD

**Issue by wdconinc** [AIDASoft/DD4hep#994]
Maybe this is somewhat resolved by out-of-band actions (running on geometry repository git tags only), but for consistency between simulations produced at one time and reconstructions performed at a later time (without storing a full copy of the geometry), it would be useful to be able to store a geometry checksum in the simulation metadata. Reconstruction code could then compare the stored geometry checksum with the checksum determined for the geometry it has access to. Reconstruction code could then decide how to proceed (depending e.g. on user input in the form of `--allow-geometry-mismatch` flags or so).

I looked through the DD4hep source tree for functionality that addresses this use case, but couldn't find anything (but naturally hash and checksum have a fair number of false positives). It is also possible that TGeo provides something applicable, but again I was not able to find anything.

What would be in the checksum?
- Essentially anything in the TGeo tree: dimensions, materials,
- but also DD4hep concepts such as segmentations and their cellID bit fields.

**Comment by wdconinc** [comment]
I'd be happy to work on implementation, but hoping for some ideas from the devs here.

**Comment by andresailer** [comment]
This is very interesting also for validation and consistency.

After reading out the TGeo Tree, read the DetElement tree with the connected segmentations and such?

**Comment by MarkusFrankATcernch** [comment]
This should be technically relatively simple. 
The problem is to have well normalized string output for each little component in dd4hep:
DetElement, Readout, all shapes, Segmentations, Materials etc...
These are largely not present, in a form which does not suffer from rounding errors etc.

Then one could simply run through 
- the structural hierarchy and the 
- geometrical hierarchy
and feed each item in a TMD5 of root. This should lead to a unique and reproducible checksum.
- For some sort of human readable output I once did part of the excercise for debugging.
- One could also use GDML to dump parts of the geometry, but here one probably suffers from rounding problems.

**Comment by MarkusFrankATcernch** [comment]
Dear @andresailer , @wdconinc ,

could you please have a look at:
https://github.com/MarkusFrankATcernch/DD4hep/commit/b589aa2df172c5a827d490da83c4d6c3accaa0fb
This includes an extensible Detector hash-facility.
Is something like this sufficient ?

Thanks

**Comment by wdconinc** [comment]
https://github.com/MarkusFrankATcernch/DD4hep/blob/989da4a6814a5de14e1d040ae3f5cbd389433ed9/DDCore/src/plugins/DetectorChecksum.cpp#L1118
```diff
-        "Usage: -plugin DD4hep_GeometryDisplay  -arg [-arg]                      \n\n"
+        "Usage: -plugin DD4hep_DetectorChecksum  -arg [-arg]                      \n\n"
```

**Comment by MarkusFrankATcernch** [comment]
Please see: https://github.com/AIDASoft/DD4hep/pull/1028

Fields are not part of the checksum. Since they are created by factories a in retro parameterization
to create a hash is not easily possible.

**Comment by MarkusFrankATcernch** [comment]
@wdconinc Ever tries PR https://github.com/AIDASoft/DD4hep/pull/1028 ?
I would like this issue to be closed if the problem is resolved....

**Comment by MarkusFrankATcernch** [comment]
Since nobody cares about testing....
This is considered fixed.
If not, please re-open.

**Comment by wdconinc** [comment]
We're currently blocked on upgrading to 1.24 or testing much in master because we can't use podio-0.16 yet. Sorry... We're tentatively going to make the switch in two weeks when support is there.
